src/Command/ServerCommand.php

```diff
@@ -106,33 +118,38 @@ public function startup(): void
             $this->_iniPath = $m[1] . '\\' . $m[2];
         }
 
-        parent::startup();
+        $this->_welcome();
     }
 
     /**
-     * Displays a header for the shell
+     * Displays a header for the command
      *
      * @return void
      */
     protected function _welcome(): void
     {
-        $this->out();
-        $this->out(sprintf('<info>Welcome to CakePHP %s Console</info>', 'v' . Configure::version()));
-        $this->hr();
-        $this->out(sprintf('App : %s', Configure::read('App.dir')));
-        $this->out(sprintf('Path: %s', APP));
-        $this->out(sprintf('DocumentRoot: %s', $this->_documentRoot));
-        $this->out(sprintf('Ini Path: %s', $this->_iniPath));
-        $this->hr();
+        $this->io->out();
+        $this->io->out(sprintf('<info>Welcome to CakePHP %s Console</info>', 'v' . Configure::version()));
+        $this->io->hr();
+        $this->io->out(sprintf('App : %s', Configure::read('App.dir')));
+        $this->io->out(sprintf('Path: %s', APP));
+        $this->io->out(sprintf('DocumentRoot: %s', $this->_documentRoot));
+        $this->io->out(sprintf('Ini Path: %s', $this->_iniPath));
+        $this->io->hr();
     }
 
     /**
-     * Override main() to handle action
+     * Execute.
      *
-     * @return void
+     * @param \Cake\Console\Arguments $args The command arguments.
+     * @param \Cake\Console\ConsoleIo $io The console io
+     * @return null|int The exit code or null for success
      */
-    public function main(): void
+    public function execute(Arguments $args, ConsoleIo $io): ?int
     {
+        $this->io = $io;
+        $this->args = $args;
+        $this->startup();
```

Comment:
Any reason to use properties over passing arguments to `startup()`? Won't we get warnings from typecheckers that `$this->args` and `$this->io` are potentially null?

Reply:
>Any reason to use properties over passing arguments to startup()?

no reason, this method was a way to modify header of this shell 
I can change to add arguments , and I can merge  the`_welcome()` and `startup()`

>Won't we get warnings from typecheckers that $this->args and $this->io are potentially null?

since order of use is : `run()`->`execute()` ->`startup()`->`_welcome()`
we won't receive any warnings IMO

Reply:
There are warnings in the psalm output for these properties.